src/proxy/http/retry.rs

```diff
@@ -0,0 +1,227 @@
+    type Error = Svc::Error;
+    type Future = tower_retry::ResponseFuture<Policy<R, St>, Svc, Request<A>>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.0.poll_ready()
+    }
+
+    fn call(&mut self, mut req: Request<A>) -> Self::Future {
+        req.extensions_mut().insert(FirstRequestStartedAt(clock::now()));
+        self.0.call(req)
+    }
+}
+
+// === impl Policy ===
+
+impl<R, S, A, B, E> ::tower_retry::Policy<Request<A>, Response<B>, E> for Policy<R, S>
+where
+    R: Retry + Clone,
+    S: Stats + Clone,
+    A: TryClone,
+{
+    type Future = future::FutureResult<Self, ()>;
+
+    fn retry(&self, req: &Request<A>, result: Result<&Response<B>, &E>) -> Option<Self::Future> {
+        match result {
+            Ok(res) => {
+                let instant = if let Some(instant) = req.extensions().get::<FirstRequestStartedAt>() {
+                    instant
+                } else {
+                    error!("retry middleware FirstRequestStartedAt extension is missing");
+                    return None;
+                };
+                match self.0.retry(instant.0, res) {
+                    Ok(()) => {
+                        trace!("retrying request");
+                        Some(future::ok(self.clone()))
+                    },
+                    Err(NoRetry::Budget) => {
+                        self.1.incr_retry_skipped_budget();
+                        None
+                    },
+                    Err(NoRetry::Timeout) => {
+                        self.1.incr_retry_skipped_timeout();
+                        None
+                    },
+                    Err(NoRetry::Success) => None,
+                }
+            },
+            Err(_err) => {
+                trace!("cannot retry transport error");
+                None
+            }
+        }
+    }
+
+    fn clone_request(&self, req: &Request<A>) -> Option<Request<A>> {
+        if let Some(clone) = req.try_clone() {
+            trace!("cloning request");
+            Some(clone)
+        } else {
+            trace!("request could not be cloned");
+            None
+        }
+    }
+}
+
+impl<B: TryClone> TryClone for Request<B> {
+    fn try_clone(&self) -> Option<Self> {
+        if let Some(body) = self.body().try_clone() {
+            let mut clone = Request::new(body);
+            *clone.method_mut() = self.method().clone();
+            *clone.uri_mut() = self.uri().clone();
+            *clone.headers_mut() = self.headers().clone();
+
+            if let Some(ext) = self.extensions().get::<::proxy::server::Source>() {
+                clone.extensions_mut().insert(ext.clone());
+            }
+            if let Some(ext) = self.extensions().get::<::app::classify::Response>() {
+                clone.extensions_mut().insert(ext.clone());
+            }
+            if let Some(ext) = self.extensions().get::<FirstRequestStartedAt>() {
+                clone.extensions_mut().insert(ext.clone());
+            }
```

Comment:
It's unfortunate that these extensions are hardocded here. I don't think the `app` types, especially, should be relied on anywhere outside of `app`.  Perhaps we should hang this trait off of the Target type, as is done for tap::Inspect?